aws/resource_aws_kms_grant.go

```diff
@@ -0,0 +1,418 @@
+		grant, err := findKmsGrantById(conn, keyId, grantId, nil)
+		if err != nil {
+			if _, ok := err.(KmsGrantMissingError); ok {
+				if grant == nil {
+					return nil
+				}
+			}
+		}
+
+		if grant != nil {
+			// Force a retry if the grant still exists
+			return resource.RetryableError(
+				fmt.Errorf("[DEBUG] Grant still exists while expected to be revoked, retyring revocation check: %s", *grant.GrantId))
+		}
+
+		return resource.NonRetryableError(err)
+	})
+
+	return err
+}
+
+// The ListGrants API defaults to listing only 50 grants
+// Use a marker to iterate over all grants in "pages"
+// NB: This function only retries on AWS Errors
+func findKmsGrantById(conn *kms.KMS, keyId string, grantId string, marker *string) (*kms.GrantListEntry, error) {
+
+	input := kms.ListGrantsInput{
+		KeyId:  aws.String(keyId),
+		Limit:  aws.Int64(int64(100)),
+		Marker: marker,
+	}
+
+	var out *kms.ListGrantsResponse
+	var err error
+	var grant *kms.GrantListEntry
+
+	err = resource.Retry(2*time.Minute, func() *resource.RetryError {
+		out, err = conn.ListGrants(&input)
+
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok {
+				if awsErr.Code() == "NotFoundException" ||
+					awsErr.Code() == "DependencyTimeoutException" ||
+					awsErr.Code() == "InternalException" {
+					return resource.RetryableError(err)
+				}
+			}
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	grant = getKmsGrantById(out.Grants, grantId)
+	if grant != nil {
+		return grant, nil
+	}
+	if *out.Truncated {
+		log.Printf("[DEBUG] KMS Grant list truncated, getting next page via marker: %s", *out.NextMarker)
+		return findKmsGrantById(conn, keyId, grantId, out.NextMarker)
+	}
+
+	return nil, NewKmsGrantMissingError(fmt.Sprintf("[DEBUG] Grant %s not found for key id: %s", grantId, keyId))
+}
+
+func expandKmsGrantConstraints(configured []interface{}) *kms.GrantConstraints {
+	if len(configured) < 1 {
+		return nil
+	}
+
+	var constraint kms.GrantConstraints
+
+	for _, raw := range configured {
+		data := raw.(map[string]interface{})
+		if contextEq, ok := data["encryption_context_equals"]; ok {
+			constraint.EncryptionContextEquals = stringMapToPointers(contextEq.(map[string]interface{}))
+		}
+		if contextSub, ok := data["encryption_context_subset"]; ok {
+			constraint.SetEncryptionContextSubset(stringMapToPointers(contextSub.(map[string]interface{})))
+		}
+	}
+
+	return &constraint
+}
+
+func flattenKmsGrantConstraints(constraint *kms.GrantConstraints) []interface{} {
+	if constraint == nil {
+		return []interface{}{}
+	}
+
+	m := make(map[string]interface{}, 0)
+
+	if constraint.EncryptionContextEquals != nil {
+		m["encryption_context_equals"] = pointersMapToStringList(constraint.EncryptionContextEquals)
+	}
+	if constraint.EncryptionContextSubset != nil {
+		m["encryption_context_subset"] = pointersMapToStringList(constraint.EncryptionContextSubset)
+	}
+
+	return []interface{}{m}
```

Comment:
Can we only have one constraint? Why does the expand version loop over it then?

Reply:
Only one of encryption_context_subset or encrytpion_context_equals can be set, but that itself can be a map. Will have to refactor this as part of the move to sets.